Added solution for Vectra Cloud

dhruvilbhatt-crest opened the pull request:
Required items, please complete
   
   Change(s):
   - Added Vectra Cloud Solution for certification

   Reason for Change(s):
   - New Product

   Version Updated:
   - Initial Version

   Testing Completed:
   - Yes

   Checked that the validations are passing and have addressed any issues that are present:
   - Yes

oshezaf reviewed (dismissed):
Before we start reviewing the ASIM parsers, please follow the following [submission guidelines](https://learn.microsoft.com/en-us/azure/sentinel/normalization-develop-parsers#contribute-parsers) specific to ASIM parsers. 

In addition, one thing I already notes: please avoid parse_json, in the ASIM parsers and in general, and use parse-kv instead.

dhruvilbhatt-crest commented:
@microsoft-github-policy-service agree company="Vectra"

devikamehra commented on Parsers/ASimAuditEvent/Parsers/ASimAuditEventVectraCloudAudit.yaml:
```diff
@@ -0,0 +1,52 @@
+          ActorUsernameType = "UPN",
+          EventResult = case(result_status_s=="success", "Success", result_status_s=="failure", "Failure","NA")
+    | project-rename
+          Dvc = source_ip_s,
```
Are we sure that source_ip_s is not the IP of the actor and is the Dvc IP Address?
If yes, DvcIpAddr should be populated and Dvc should be the Alias

dhruvilbhatt-crest replied:
@devikamehra We have mapped the source_ip_s with the DvcIpAddr field and additionally, we have also added Dvc column that corresponds to the DvcIpAddr field.

devikamehra commented on Parsers/ASimAuditEvent/Parsers/ASimAuditEventVectraCloudAudit.yaml:
```diff
@@ -0,0 +1,52 @@
+          EventProductVersion = version_s
+    | extend User = ActorUsername
+    | project-away
+          id_d, user_id_d, user_role_s, result_status_s,event_timestamp_t, event_data_s, api_client_id_g, TenantId, _ResourceId, _SubscriptionId, RawData, SourceSystem, Computer, MG, ManagementGroupName
```
Use project-away *_s instead.

dhruvilbhatt-crest replied:
@devikamehra We have updated the KQL queries in all the ASIM files.

devikamehra commented on Parsers/ASimAuditEvent/Parsers/ASimAuditEventVectraCloudAudit.yaml:
```diff
@@ -0,0 +1,52 @@
+          EventStartTime = event_timestamp_t,
+          EventType = "Other",
+          EventVendor = 'Vectra',
+          Type = "Audit Log",
```
Type field is auto-populated and need not be populated again.

v-prasadboke commented:
Hello @dhruvilbhatt-crest please go through the above comment and clear merge conflicts

dhruvilbhatt-crest commented on Parsers/ASimAuditEvent/Parsers/ASimAuditEventVectraCloudAccountDetection.yaml:
```diff
@@ -0,0 +1,57 @@
+Parser:
```
Hi @devikamehra , 

We mapped the detection data with the AuditEvent Schema as that's the closest match we found. Though the detection data is not Audit data, it will be continuously updating based on the detection modification. 

If not AuditEvent Schema, what schema would you suggest us to map detection data with?

devikamehra replied:
@dhruvilbhatt-crest Currently we do not have any schema to map these detections. For alerts, we usually create Alert/Incidents using Analytical Rules and thus SecurityAlert table is the source of truth.
For mapping dynamic fields, Alert Details can be enriched. 
Source: https://learn.microsoft.com/en-us/azure/sentinel/customize-alert-details

I understand that this might not be exactly what you were looking for and thus happy to discuss this further.

dhruvilbhatt-crest replied:
Hi @devikamehra, we have removed the Account Detection data mapping with AuditEvent Schema.

dhruvilbhatt-crest commented on Parsers/ASimAuditEvent/Parsers/ASimAuditEventVectraCloudAudit.yaml:
```diff
@@ -0,0 +1,52 @@
+  Link: https://aka.ms/AboutASIM
+Description: |
+  This ASIM parser supports normalizing VectraCloud Audit Logs Event in the Audits_Data_CL table to the ASIM Audit Event schema.
+ParserName: ASimAuditEventVectraCloudAudit
```
@devikamehra Added the EquivalentBuiltInParser key-value pair in all the ASIM files.

dhruvilbhatt-crest commented on Parsers/ASimAuditEvent/Parsers/ASimAuditEventVectraCloudAudit.yaml:
```diff
@@ -0,0 +1,52 @@
+Product:
+  Name: Vectra
+Normalization:
+  Schema: AuditEvent
```
@devikamehra Vectra Audit data has additional data along with login/logout messages. We have included additional sample data for the Vectra Audit. This should give you a better understanding of the Vectra Audit data.

devikamehra replied:
@dhruvilbhatt-crest I still see only login events in the sample data - https://github.com/Azure/Azure-Sentinel/blob/aa8c2cceb4865cefd007be9a54775e0f5a6a7f82/Sample%20Data/Custom/VectraAudits.csv

Can you please guide me where the updated logs are?

Also, in case we have more than one type of data in the Audit logs, we can filter the data it in the parser and develop Authentication ASim Parser for authentication while AuditEvents ASim Parser for the rest.

dhruvilbhatt-crest replied:
Hi @devikamehra, 

You can find the ASim sample data here: https://github.com/dhruvilbhatt-crest/Azure-Sentinel/tree/vectra-cloud/Sample%20Data/ASIM/VectraCloud

Per our understanding, the data will belong to Audit Events only. Upon reviewing the sample data available at the link above, we may discuss further.

devikamehra replied:
@dhruvilbhatt-crest Thank you for providing additional information. Wanted to confirm if we only get login success  events in this table or there can be logout or failures as well? If yes, in my opinion, it makes total sense to have an authentication parser for this table filtered on event_type.

For all the remaining event_types, Audit Schema is more appropriate and we should leverage object and operation fields for each event_type effectively.

dhruvilbhatt-crest replied:
Hi @devikamehra, we have introduced Authentication Parser and made the necessary changes.

dhruvilbhatt-crest commented on Parsers/ASimAuditEvent/Parsers/ASimAuditEventVectraCloudAudit.yaml:
```diff
@@ -0,0 +1,52 @@
+          EventType = "Other",
+          EventVendor = 'Vectra',
+          Type = "Audit Log",
+          EventUid = tostring(toint(id_d)),
```
@devikamehra We noticed that when converting the "real" datatype columns to "string", the float values were being generated. For example, while converting 'id_d' with a value of 1, we were getting '1.0'. 

In accordance with your suggestion, we have removed the toint() function. However, if you believe it is necessary, please let us know and we will reinstate it.

devikamehra replied:
Understood the concern. If we are getting '1.0' as the result, it would be better to use toint().

dhruvilbhatt-crest commented on Parsers/ASimAuditEvent/Parsers/ASimAuditEventVectraCloudAudit.yaml:
```diff
@@ -0,0 +1,52 @@
+          Type = "Audit Log",
+          EventUid = tostring(toint(id_d)),
+          ActorUserId = tostring(toint(user_id_d)),
+          ActorUserIdType = "UID",
```
@devikamehra We wanted to bring to your attention that the Microsoft Sentinel ASIM has a limited number of values that can be used for the ActorUserIdType column. Ref: https://learn.microsoft.com/en-us/azure/sentinel/normalization-schema-audit#actor-fields. Based on our analysis of the Vectra Audit data, we have found that the most suitable ActorUserIdType to use is "UID". However, when we tried to add a new ActorUserIdType as per your suggestion, we encountered an error during the testing of the ASIM parser. Specifically, the error message stated that there was 1 invalid value(s) (up to 10 listed) in 100 records (100.0%) for the field [ActorUserIdType] of type [Enumerated], with the value "VectraUserId" (Schema:AuditEvent).

devikamehra replied:
Please proceed with VectraUserId and I will take care of adding this as a valid ActorUserIdType.

v-prasadboke commented:
Hello @devikamehra and @oshezaf author has made the changes

devikamehra commented on Parsers/ASimAuthentication/Parsers/ASimAuthenticationVectraCloudAudit.yaml:
```diff
@@ -0,0 +1,54 @@
+          EventSchema = "Authentication",
+          EventSchemaVersion = "0.1.3",
+          EventStartTime = event_timestamp_t,
+          EventType = case(event_action_s=="login", "Logon", event_action_s=="logout", "Logoff","Elevate"),
```
Since only two types of event_action_s fields exist, default should be an empty string.

dhruvilbhatt-crest replied:
@devikamehra Updated.

devikamehra commented on Parsers/ASimAuthentication/Parsers/ASimAuthenticationVectraCloudAudit.yaml:
```diff
@@ -0,0 +1,54 @@
+          ActorUserId = tostring(toint(user_id_d)),
+          ActorUserIdType = "VectraUserId",
+          ActorUsernameType = "UPN",
+          EventUid = tostring(toint(id_d))
```
Just a confirmation required, is api_client_id_g a unique id for this event or id_d is sufficient to identify this event uniquely?

dhruvilbhatt-crest replied:
@devikamehra id_d is the unique field for each event. api_client_id_g is just another field with very limited occurrences.

devikamehra commented on Parsers/ASimAuthentication/Parsers/ASimAuthenticationVectraCloudAudit.yaml:
```diff
@@ -0,0 +1,54 @@
+          ActorUsernameType = "UPN",
+          EventUid = tostring(toint(id_d))
+    | project-rename
+          DvcIpAddr = source_ip_s,
```
Is this IP of the source user or the device creating this event?

dhruvilbhatt-crest replied:
@devikamehra The source_ip_s will contain the IP address of the device from which the event is generated.

v-prasadboke commented:
Hello @devikamehra any updates on this

v-prasadboke commented:
Hello @devikamehra waiting for your updates

devikamehra commented on Parsers/ASimAuditEvent/Parsers/ASimAuditEventVectraCloudAudit.yaml:
```diff
@@ -0,0 +1,54 @@
+  let parser = (disabled:bool = false)
+  {
+    Audits_Data_CL
+    | extend
```
Just how we added filters in Authentication, we would have to remove that data from this parser. Otherwise same data would get parsed here as well.

dhruvilbhatt-crest replied:
Addressed.

devikamehra commented on Parsers/ASimAuditEvent/Parsers/vimAuditEventVectraCloudAudit.yaml:
```diff
@@ -0,0 +1,74 @@
+          ActorOriginalUserType = user_type_s,
+          EventMessage = Message,
+          EventProductVersion = version_s
+    | where ('*' in (eventresult) or EventResult in (eventresult))
```
eventresult is a string
We should have eventresult == "*" or EventResult == eventresult

dhruvilbhatt-crest replied:
Updated.

devikamehra commented on Parsers/ASimAuditEvent/Parsers/vimAuditEventVectraCloudAudit.yaml:
```diff
@@ -0,0 +1,74 @@
+  let parser = (disabled:bool = false, eventresult:string='*', starttime:datetime=datetime(null), endtime:datetime=datetime(null), actorusername_has_any:dynamic=dynamic([]),operation_has_any:dynamic=dynamic([]),object_has_any:dynamic=dynamic([]))
+  {
+    Audits_Data_CL
+    | where (isnull(starttime) or event_timestamp_t >= starttime) and (isnull(endtime) or event_timestamp_t <= endtime) and (array_length(actorusername_has_any) == 0 or tostring(toint(user_id_d)) has_any (actorusername_has_any)) or (array_length(actorusername_has_any) == 0 or username_s has_any (actorusername_has_any)) and (array_length(operation_has_any) == 0 or event_action_s has_any (operation ...
```
Filter for disabled is missing for all the parsers

dhruvilbhatt-crest replied:
Updated.

v-prasadboke commented:
Hello @dhruvilbhatt-crest please remove branch conflicts as well

dhruvilbhatt-crest commented:
@v-prasadboke @devikamehra All the comments have been addressed and the merge conflict is resolved.

v-prasadboke commented:
Hello @devikamehra waiting for your approval

devikamehra commented:
@v-prasadboke I am working on adding required fields to test scenarios and will approve post that.
@dhruvilbhatt-crest Not necessary, but a suggestion. Can we add EventSeverity based on EventResult field.

v-prasadboke commented:
Thank you @devikamehra

dhruvilbhatt-crest commented:
Hi @devikamehra , 

Thank you for your suggestion, but we discussed it internally, and we didn't find this useful. We will keep this in mind in the future.

dhruvilbhatt-crest commented:
@devikamehra @v-prasadboke All the comments have been addressed. Do we have any updates on this?

v-prasadboke commented on Solutions/Vectra Cloud/Monitor Alert Rules/Alert_When_Function_app_Fails.json:
Hello @dhruvilbhatt-crest is thile file a detection, If yes please rename the folder name as Analytic Rules

dhruvilbhatt-crest replied:
@v-prasadboke - This is not the Analytic Rule but the "Monitor Alert Rule". Do you still want us to add that under the Analytic Rules folder or do you have any specific directory for that?

v-prasadboke commented on Solutions/Vectra Cloud/Parsers/VectraAccountDetection.txt:
And for this parser, Its now showing any data for the table Sample Account_Detection_Data_CL.csv after ingesting the data.

dhruvilbhatt-crest replied:
We have provided the sample data in the path: Sample Data/Custom/Account_Detection_Data_CL.csv

You can create the table name Account_Detection_Data_CL from the sample data provided at the above path, and to use you can configure the custom parser with the name as VectraAccountDetection which refers to the custom table name Account_Detection_Data_CL which is created from the above-provided sample data.

Note: We have created the sample data files with the custom table names which are further referenced in the Custom Parsers as well.
Let me know if we need to update the sample data file name or the parser or maybe the table name referenced in the custom parser.

devikamehra commented:
@dhruvilbhatt-crest There is a PR pending to be merged for whitelisting the required strings in tests
https://github.com/Azure/Azure-Sentinel/pull/8047

dhruvilbhatt-crest commented:
@devikamehra Since the PR [#8047](https://github.com/Azure/Azure-Sentinel/pull/8047) is merged, can you approve this if nothing is pending for review?

dhruvilbhatt-crest commented:
@v-prasadboke @devikamehra Any updates on this PR?

dhruvilbhatt-crest commented:
Hi @v-prasadboke @devikamehra, 

Can we have some updates here?

devikamehra reviewed (dismissed):
Validated ASIM Parsers only.
Please merge if everything else is good.

dhruvilbhatt-crest commented:
Thanks for confirming Devika. 

@v-prasadboke Can you provide approval if it's reviewed?

v-prasadboke commented:
Hello @dhruvilbhatt-crest ,Just to make sure the workbook images you have added are working right??

And Please update the api version of maintemplate.
After the updating the api version please update the zip with the latest create ui and  maintemplate, This is the only change requested other than this everything looks good.

dhruvilbhatt-crest commented:
@v-prasadboke Yes, the images are working. Also, the API version has been updated in the maintemplate and zip file.

dhruvilbhatt-crest commented:
@v-prasadboke Are we good to merge this PR? Customers are eagerly waiting on this.

v-prasadboke commented:
> @v-prasadboke Are we good to merge this PR? Customers are eagerly waiting on this.

Hello @dhruvilbhatt-crest looks like you still havent updated the API version, please update the api version of Microsoft.OperationalInsights/workspaces/savedSearches which is 2021-06-01 and replace it with 2022-10-01

v-prasadboke commented:
After updating the main template please update the zip with the latest maintemplate and createui

v-prasadboke commented:
Hello @dhruvilbhatt-crest please update the api version in maintemplate

dhruvilbhatt-crest commented:
@v-prasadboke Updated the API version. Let me know if we are good to merge this or what all changes are required to get this merged.

v-prasadboke commented:
Hello @dhruvilbhatt-crest please add the workbook metadata in the ' workbookmetadata ' file in the given path.
https://github.com/Azure/Azure-Sentinel/blob/master/Tools/Create-Azure-Sentinel-Solution/V2/WorkbookMetadata/WorkbooksMetadata.json

dhruvilbhatt-crest commented:
@v-prasadboke Updated the workbook metadata.

dhruvilbhatt-crest commented:
@v-prasadboke Any update on the PR?

v-prasadboke commented:
Hello @dhruvilbhatt-crest we are good to go just remove branch conflicts

dhruvilbhatt-crest commented:
@v-prasadboke Resolved the branch conflicts.

v-prasadboke commented:
All looks good

dhruvilbhatt-crest commented:
@v-prasadboke @devikamehra @oshezaf - Please hold off merging this PR as of now. While this PR was in review, we were working on adding some new features and it will be ready to merge in a week. 

Affected areas:
- Data Connectors
- Workbooks

Note: There are **NOT** going to be any changes in ASIM parsers and Custom Parsers.

dhruvilbhatt-crest commented:
@v-prasadboke @oshezaf @devikamehra 

Updated the PR with the following:

1. Re-branding the app from "Vectra Cloud" to "Vectra XDR"
2. Added data connectors and workbooks for "Lockdown" and "Health".
3. Updated existing data connectors and workbooks.

NOTE: No changes have been done in the ASIM parsers apart from re-branding.

Commit link: https://github.com/Azure/Azure-Sentinel/pull/7748/commits/984f411b88d80952eb39216d1fae602321b923a1

v-prasadboke commented:
Hello @dhruvilbhatt-crest will be investigating this PR, you can expect an eta of 14 Jul

v-prasadboke commented:
Hello @dhruvilbhatt-crest there are still validation failure for codeql and the PR has still Merge conflicts, please resolve the issue

v-prasadboke commented:
The folder structure is a bit incorrect, Please remove workbook metadata from the workbook folder of solution. 
Workbook folder should only contain workbook and workbook images in a folder.
And images from Workbook individual folder  should be removed as well.

v-prasadboke commented on Solutions/Vectra XDR/SolutionMetadata.json:
This should not be included in the folder

v-prasadboke commented on Workbooks/Images/Preview/VectraXDRBlack1.png:
please remove all the images from this workbook folder as they already exist in the solution folder

v-prasadboke commented:
Hello @dhruvilbhatt-crest any updates on the above

dhruvilbhatt-crest commented:
Hi @v-prasadboke, 

Sorry for the delay, but we are working on some additional changes in this PR. I will fix your review comments along with the changes soon.

v-prasadboke commented:
Ok @dhruvilbhatt-crest thanks for the update

v-prasadboke commented:
Hello @dhruvilbhatt-crest Good Evening, we know you are still working on this solution and adding efforts to the fullest.
But if possible can we have any update on this PR.

Thank you,
Prasad boke.

v-prasadboke commented:
Hi @dhruvilbhatt-crest, We are looking forward to hearing from you regarding the following comments.

v-prasadboke commented:
Hello @dhruvilbhatt-crest, please provide some updates.

dhruvilbhatt-crest commented:
Hi @v-prasadboke , 

Sorry for the delayed response. We are almost done with the changes and currently working on the conflict. We will push the final changes within few hours.

v-prasadboke commented:
Thank you @dhruvilbhatt-crest for the updates.

dhruvilbhatt-crest commented:
@v-prasadboke @devikamehra Thanks for waiting on this. We have pushed all the changes and this PR is now good to review.

v-prasadboke commented:
Thank you @dhruvilbhatt-crest for committing the changes, You can expect an update till 01 Aug 2023.

v-prasadboke commented:
Hello @dhruvilbhatt-crest, All looks good.
But can you please repackage the solution using V3 Tool and Release Notes as well.

Instructions to use V3 : https://github.com/Azure/Azure-Sentinel/blob/v-prasadboke-sampledata/Tools/Create-Azure-Sentinel-Solution/V3/README.md

Instructions for Release notes: https://github.com/Azure/Azure-Sentinel/blob/master/Solutions/ReleaseNotesGuidance.md
Thank you.

dhruvilbhatt-crest commented:
@v-prasadboke Added the ReleaseNotes.md file

dhruvilbhatt-crest commented:
@v-atulyadav @v-prasadboke - While the client tried to install the published solution of Vectra XDR for Microsoft Sentinel on their tenant, they are getting an error as shown below:

Can you please help us understand the error, possible root cause, and solution for the same?

![image](https://github.com/Azure/Azure-Sentinel/assets/109515601/6d1abb5b-2755-42ad-b3f5-44a82029c287)

v-prasadboke commented:
Hello @dhruvilbhatt-crest, Thank you bringing this to our attention. We will look into this and inform you about the same soonest.

v-prasadboke commented:
Hello @dhruvilbhatt-crest, RCA for this issue is while you repackaged the solution the workbook metadata which existed in the file workbookmetadata of V2 folder was duplicated. That might be the reason while you repackaged the solution due to overlapping of workbook metadata the tool was unable to pull the metadata of workbook from Workbookmetadata file of V2 folder.
While testing the solution in my local, the solution was deployed successfully without any errors.
<img width="681" alt="image" src="https://github.com/Azure/Azure-Sentinel/assets/117061676/88907d86-1d84-4900-a0f0-ebafc2e2a726">

v-prasadboke commented:
I've even raised a PR for the same.
Link to the PR :-  [8832](https://github.com/Azure/Azure-Sentinel/pull/8832)

dhruvilbhatt-crest commented:
Thanks @v-prasadboke for looking into this. Is there any action item in my bucket?

v-prasadboke commented:
Hello @dhruvilbhatt-crest, No not much for you to work on. If needed I'll inform you for such.

dhruvilbhatt-crest commented:
Hi @v-prasadboke , 

While clicking on the "Deploy to Azure" button with v3.0.1, redirection was not working while configuring the data connector. It's redirecting to the "Bing" search page. 

I have raised another PR after renaming the Data connector zip file and fixing the zip path. Please review and let us know if there could be any other possible reason for this redirection. 

PR Link: https://github.com/Azure/Azure-Sentinel/pull/8859
